File: tests/test_e2e.js
/**
 * E2E Tests for IDS2PSET - Full cycle testing
 */

import { describe, it, beforeEach, afterEach } from 'node:test';
import assert from 'node:assert';
import { readFileSync, writeFileSync, unlinkSync, existsSync, readdirSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const FIXTURES_DIR = join(__dirname, 'fixtures');
const OUTPUT_DIR = join(__dirname, 'output');

// Helper to check if file exists
function fileExists(path) {
    try {
        return existsSync(path);
    } catch {
        return false;
    }
}

describe('E2E: IDS2PSET Full Cycle', () => {
    beforeEach(() => {
        // Create output directory if not exists
        if (!fileExists(OUTPUT_DIR)) {
            // Note: Can't create dirs in node:test without additional setup
        }
    });

    afterEach(() => {
        // Cleanup output files
        const outputFiles = [
            join(OUTPUT_DIR, 'test_output.ifc')
        ];
        outputFiles.forEach(file => {
            if (fileExists(file)) {
                try {
                    unlinkSync(file);
                } catch (e) {
                    // Ignore cleanup errors
                }
            }
        });
    });

    describe('Fixture Files', () => {
        it('should have minimal.ids fixture', () => {
            const fixturePath = join(FIXTURES_DIR, 'minimal.ids');
            assert.ok(fileExists(fixturePath), 'minimal.ids should exist');

            const content = readFileSync(fixturePath, 'utf-8');
            assert.ok(content.includes('<?xml'), 'Should be valid XML');
            assert.ok(content.includes('<ids'), 'Should be IDS format');
        });

        it('should have with_enumeration.ids fixture', () => {
            const fixturePath = join(FIXTURES_DIR, 'with_enumeration.ids');
            assert.ok(fileExists(fixturePath), 'with_enumeration.ids should exist');

            const content = readFileSync(fixturePath, 'utf-8');
            assert.ok(content.includes('enumeration'), 'Should contain enumeration');
        });

        it('should have multi_entity.ids fixture', () => {
            const fixturePath = join(FIXTURES_DIR, 'multi_entity.ids');
            assert.ok(fileExists(fixturePath), 'multi_entity.ids should exist');

            const content = readFileSync(fixturePath, 'utf-8');
            assert.ok(content.includes('IFCWALL'), 'Should contain IFCWALL');
            assert.ok(content.includes('IFCSLAB'), 'Should contain IFCSLAB');
        });
    });

    describe('Application Structure', () => {
        it('should have index.html', () => {
            const indexPath = join(__dirname, '..', 'index.html');
            assert.ok(fileExists(indexPath), 'index.html should exist');

            const content = readFileSync(indexPath, 'utf-8');
            assert.ok(content.includes('IDS2PSET'), 'Should have app title');
            assert.ok(content.includes('drop-zone'), 'Should have drop zone');
            assert.ok(content.includes('download-btn'), 'Should have download button');
        });

        it('should have css/style.css', () => {
            const cssPath = join(__dirname, '..', 'css', 'style.css');
            assert.ok(fileExists(cssPath), 'style.css should exist');

            const content = readFileSync(cssPath, 'utf-8');
            assert.ok(content.includes('.drop-zone'), 'Should style drop zone');
            assert.ok(content.includes('.btn'), 'Should style buttons');
        });

        it('should have js/app.js', () => {
            const appPath = join(__dirname, '..', 'js', 'app.js');
            assert.ok(fileExists(appPath), 'app.js should exist');

            const content = readFileSync(appPath, 'utf-8');
            assert.ok(content.includes('IDS2PSETApp'), 'Should have main class');
            assert.ok(content.includes('download'), 'Should have download method');
            assert.ok(content.includes('generateIFC'), 'Should generate IFC');
        });

        it('should have js/pyodide-bridge.js', () => {
            const bridgePath = join(__dirname, '..', 'js', 'pyodide-bridge.js');
            assert.ok(fileExists(bridgePath), 'pyodide-bridge.js should exist');

            const content = readFileSync(bridgePath, 'utf-8');
            assert.ok(content.includes('PyodideBridge'), 'Should have bridge class');
            assert.ok(content.includes('parseIDS'), 'Should parse IDS');
            assert.ok(content.includes('generateIFC'), 'Should generate IFC');
            assert.ok(content.includes('validateIFC'), 'Should validate IFC');
        });
    });

    describe('Python Modules', () => {
        it('should have src/ids_parser.py', () => {
            const parserPath = join(__dirname, '..', 'src', 'ids_parser.py');
            assert.ok(fileExists(parserPath), 'ids_parser.py should exist');

            const content = readFileSync(parserPath, 'utf-8');
            assert.ok(content.includes('parse_ids_file'), 'Should have parse function');
            assert.ok(content.includes('PSetGroup'), 'Should have PSetGroup class');
        });

        it('should have src/pset_generator.py', () => {
            const genPath = join(__dirname, '..', 'src', 'pset_generator.py');
            assert.ok(fileExists(genPath), 'pset_generator.py should exist');

            const content = readFileSync(genPath, 'utf-8');
            assert.ok(content.includes('PSetGenerator'), 'Should have generator class');
            assert.ok(content.includes('IfcProjectLibrary'), 'Should create library');
        });

        it('should have src/validator.py', () => {
            const valPath = join(__dirname, '..', 'src', 'validator.py');
            assert.ok(fileExists(valPath), 'validator.py should exist');

            const content = readFileSync(valPath, 'utf-8');
            assert.ok(content.includes('IFCValidator'), 'Should have validator class');
            assert.ok(content.includes('validate_file'), 'Should validate files');
        });

        it('should have src/gherkin_validator.py', () => {
            const gherkinPath = join(__dirname, '..', 'src', 'gherkin_validator.py');
            assert.ok(fileExists(gherkinPath), 'gherkin_validator.py should exist');

            const content = readFileSync(gherkinPath, 'utf-8');
            assert.ok(content.includes('GherkinValidator'), 'Should have validator class');
            assert.ok(content.includes('RULE_CATEGORIES'), 'Should have rule categories');
            assert.ok(content.includes('validate_all_rules'), 'Should validate all rules');
        });
    });

    describe('Wheel Files', () => {
        it('should have ifcopenshell wheel', () => {
            const wheelDir = join(__dirname, '..', 'wheels');
            assert.ok(fileExists(wheelDir), 'wheels directory should exist');

            // Check for any .whl file
            const files = readdirSync(wheelDir);
            const wheelFiles = files.filter(f => f.endsWith('.whl'));
            assert.ok(wheelFiles.length > 0, 'Should have at least one wheel file');
        });
    });

    describe('Git Submodules', () => {
        it('should have ifc-gherkin-rules submodule', () => {
            const submodulePath = join(__dirname, '..', 'vendor', 'ifc-gherkin-rules');
            assert.ok(fileExists(submodulePath), 'ifc-gherkin-rules submodule should exist');

            const mainPy = join(submodulePath, 'main.py');
            assert.ok(fileExists(mainPy), 'Should have main.py');
        });
    });
});

describe('E2E: Integration Flow', () => {
    it('should have complete data flow documented', () => {
        // This test verifies the integration flow is documented
        const flow = [
            '1. User uploads IDS file via drag-and-drop',
            '2. File content read via File API',
            '3. Pyodide loads Python modules',
            '4. ids_parser.py parses IDS XML',
            '5. PSet data grouped by propertySet name',
            '6. UI displays PSet tree with checkboxes',
            '7. User selects PSet to include',
            '8. pset_generator.py generates IFC4',
            '9. validator.py validates IFC structure',
            '10. gherkin_validator.py checks buildingSMART rules',
            '11. IFC blob created and downloaded'
        ];

        assert.ok(flow.length === 11, 'Should have complete flow documented');
    });
});

File: tests/test_e2e.py
"""E2E Tests for IDS2PSET - Full cycle testing."""

import pytest
import sys
import os

# Add src to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

try:
    from ids_parser import parse_ids_file
    from pset_generator import PSetGenerator
    from validator import IFCValidator
    from gherkin_validator import GherkinValidator, GHERKIN_AVAILABLE

    ALL_MODULES_AVAILABLE = True
except ImportError:
    ALL_MODULES_AVAILABLE = False


@pytest.mark.skipif(not ALL_MODULES_AVAILABLE, reason="Modules not available")
class TestE2EFullCycle:
    """End-to-End tests for complete IDS → IFC workflow."""

    def test_minimal_ids_full_cycle(self):
        """Test complete cycle with minimal IDS fixture."""
        fixture_path = os.path.join(
            os.path.dirname(__file__), "fixtures", "minimal.ids"
        )

        # Step 1: Parse IDS
        parsed = parse_ids_file(fixture_path)
        assert len(parsed) > 0, "Should parse at least one PSet"

        # Step 2: Convert to generator format
        psets_data = {}
        for name, pset in parsed.items():
            psets_data[name] = {
                "name": pset.name,
                "properties": [
                    {
                        "name": p.name,
                        "data_type": p.data_type,
                        "description": p.description,
                        "enum_values": p.enum_values,
                        "cardinality": p.cardinality,
                    }
                    for p in pset.properties
                ],
                "applicable_entities": pset.applicable_entities,
            }

        # Step 3: Generate IFC
        generator = PSetGenerator()
        ifc_content = generator.generate(psets_data)
        assert len(ifc_content) > 0, "Should generate IFC content"
        assert "IFC4" in ifc_content, "Should be IFC4 format"
        assert "IFCPROJECTLIBRARY" in ifc_content, "Should have library"

        # Step 4: Validate with ifcopenshell
        validator = IFCValidator()
        result = validator.validate_string(ifc_content)
        assert result["valid"], f"Should be valid IFC: {result['errors']}"

        # Step 5: Validate with gherkin rules
        if GHERKIN_AVAILABLE:
            gherkin_validator = GherkinValidator()
            gherkin_result = gherkin_validator.validate_string(ifc_content)
            assert gherkin_result is not None, "Gherkin validation should complete"

    def test_enumeration_ids_full_cycle(self):
        """Test complete cycle with enumeration IDS fixture."""
        fixture_path = os.path.join(
            os.path.dirname(__file__), "fixtures", "with_enumeration.ids"
        )

        # Parse
        parsed = parse_ids_file(fixture_path)
        assert "MaterialPSet" in parsed, "Should have MaterialPSet"

        # Convert
        psets_data = {}
        for name, pset in parsed.items():
            psets_data[name] = {
                "name": pset.name,
                "properties": [
                    {
                        "name": p.name,
                        "data_type": p.data_type,
                        "description": p.description,
                        "enum_values": p.enum_values,
                        "cardinality": p.cardinality,
                    }
                    for p in pset.properties
                ],
                "applicable_entities": pset.applicable_entities,
            }

        # Generate
        generator = PSetGenerator()
        ifc_content = generator.generate(psets_data)

        # Validate
        validator = IFCValidator()
        result = validator.validate_string(ifc_content)
        assert result["valid"], f"Should be valid IFC: {result['errors']}"

    def test_multi_entity_full_cycle(self):
        """Test complete cycle with multi-entity IDS fixture."""
        fixture_path = os.path.join(
            os.path.dirname(__file__), "fixtures", "multi_entity.ids"
        )

        # Parse
        parsed = parse_ids_file(fixture_path)
        assert "CommonPSet" in parsed, "Should have CommonPSet"

        # Verify entities are merged
        pset = parsed["CommonPSet"]
        assert "IFCWALL" in pset.applicable_entities
        assert "IFCSLAB" in pset.applicable_entities

        # Convert
        psets_data = {
            "CommonPSet": {
                "name": pset.name,
                "properties": [
                    {
                        "name": p.name,
                        "data_type": p.data_type,
                        "description": p.description,
                        "enum_values": p.enum_values,
                        "cardinality": p.cardinality,
                    }
                    for p in pset.properties
                ],
                "applicable_entities": pset.applicable_entities,
            }
        }

        # Generate
        generator = PSetGenerator()
        ifc_content = generator.generate(psets_data)

        # Validate
        validator = IFCValidator()
        result = validator.validate_string(ifc_content)
        assert result["valid"], f"Should be valid IFC: {result['errors']}"

    def test_pset_selection_filtering(self):
        """Test filtering PSet by selection."""
        fixture_path = os.path.join(
            os.path.dirname(__file__), "fixtures", "multi_entity.ids"
        )

        # Parse
        parsed = parse_ids_file(fixture_path)

        # Convert
        psets_data = {}
        for name, pset in parsed.items():
            psets_data[name] = {
                "name": pset.name,
                "properties": [
                    {
                        "name": p.name,
                        "data_type": p.data_type,
                        "description": p.description,
                        "enum_values": p.enum_values,
                        "cardinality": p.cardinality,
                    }
                    for p in pset.properties
                ],
                "applicable_entities": pset.applicable_entities,
            }

        # Select only specific PSet
        selected_names = list(psets_data.keys())[:1]
        filtered_psets = {
            name: pset for name, pset in psets_data.items() if name in selected_names
        }

        # Generate
        generator = PSetGenerator()
        ifc_content = generator.generate(filtered_psets)

        # Validate
        validator = IFCValidator()
        result = validator.validate_string(ifc_content)
        assert result["valid"], f"Should be valid IFC: {result['errors']}"

    def test_all_gherkin_categories(self):
        """Test validation against all 33 gherkin rule categories."""
        if not GHERKIN_AVAILABLE:
            pytest.skip("Gherkin validator not available")

        fixture_path = os.path.join(
            os.path.dirname(__file__), "fixtures", "minimal.ids"
        )

        # Parse and generate
        parsed = parse_ids_file(fixture_path)
        psets_data = {
            name: {
                "name": pset.name,
                "properties": [
                    {
                        "name": p.name,
                        "data_type": p.data_type,
                        "description": p.description,
                        "enum_values": p.enum_values,
                        "cardinality": p.cardinality,
                    }
                    for p in pset.properties
                ],
                "applicable_entities": pset.applicable_entities,
            }
            for name, pset in parsed.items()
        }

        generator = PSetGenerator()
        ifc_content = generator.generate(psets_data)

        # Validate against all rules
        validator = GherkinValidator()
        result = validator.validate_all_rules(ifc_content)

        assert result is not None
        assert "results_by_category" in result
        assert (
            len(result["results_by_category"]) == 33
        ), "Should validate all 33 categories"

    def test_ifc_content_structure(self):
        """Test generated IFC has correct structure."""
        fixture_path = os.path.join(
            os.path.dirname(__file__), "fixtures", "minimal.ids"
        )

        parsed = parse_ids_file(fixture_path)
        psets_data = {
            name: {
                "name": pset.name,
                "properties": [
                    {
                        "name": p.name,
                        "data_type": p.data_type,
                        "description": p.description,
                        "enum_values": p.enum_values,
                        "cardinality": p.cardinality,
                    }
                    for p in pset.properties
                ],
                "applicable_entities": pset.applicable_entities,
            }
            for name, pset in parsed.items()
        }

        generator = PSetGenerator()
        ifc_content = generator.generate(psets_data)

        # Check IFC structure
        assert "ISO-10303-21" in ifc_content, "Should have IFC header"
        assert "END-ISO-10303-21" in ifc_content, "Should have IFC footer"
        assert "IFCPROJECT" in ifc_content, "Should have IfcProject"
        assert "IFCPROJECTLIBRARY" in ifc_content, "Should have IfcProjectLibrary"
        assert "IFCPROPERTYSETTEMPLATE" in ifc_content, "Should have templates"
        assert (
            "IFCSIMPLEPROPERTYTEMPLATE" in ifc_content
        ), "Should have simple templates"


@pytest.mark.skipif(not ALL_MODULES_AVAILABLE, reason="Modules not available")
class TestE2EErrorHandling:
    """E2E tests for error handling."""

    def test_invalid_ids_handling(self):
        """Test handling of invalid IDS content."""
        invalid_ids = """<?xml version="1.0"?>
        <ids>
            <invalid_structure>
        </ids>"""

        # Should not crash
        import tempfile

        with tempfile.NamedTemporaryFile(mode="w", suffix=".ids", delete=False) as f:
            f.write(invalid_ids)
            temp_path = f.name

        try:
            # May raise exception or return empty result
            result = parse_ids_file(temp_path)
            assert isinstance(result, dict)
        except Exception:
            pass  # Expected for invalid XML
        finally:
            os.unlink(temp_path)

    def test_empty_pset_handling(self):
        """Test handling of empty PSet data."""
        generator = PSetGenerator()
        ifc_content = generator.generate({})

        # Should generate valid (but empty) IFC
        assert len(ifc_content) > 0
        assert "IFC4" in ifc_content


@pytest.mark.skipif(not ALL_MODULES_AVAILABLE, reason="Modules not available")
class TestIFCValidationScript:
    """Tests for validate_ifc_playwright.py script."""

    def test_validate_generated_ifc(self):
        """Test validation script with generated IFC."""
        from validate_ifc_playwright import validate_ifc_file
        import tempfile

        # Generate IFC
        generator = PSetGenerator()
        psets = {
            "TestPSet": {
                "name": "TestPSet",
                "properties": [
                    {
                        "name": "Property1",
                        "data_type": "IFCTEXT",
                        "description": "",
                        "enum_values": [],
                        "cardinality": "required",
                    }
                ],
                "applicable_entities": ["IFCWALL"],
            }
        }
        ifc_content = generator.generate(psets)

        # Write to temp file
        with tempfile.NamedTemporaryFile(mode="w", suffix=".ifc", delete=False) as f:
            f.write(ifc_content)
            temp_path = f.name

        try:
            # Run validation
            result = validate_ifc_file(temp_path)
            assert result is True, "Validation should pass"
        finally:
            os.unlink(temp_path)

    def test_validate_fixture_ifc(self):
        """Test validation script with Playwright generated IFC."""
        from validate_ifc_playwright import validate_ifc_file

        ifc_path = os.path.join(
            os.path.dirname(__file__),
            "..",
            ".playwright-mcp",
            "IDS2PSET-Library.ifc",
        )

        if not os.path.exists(ifc_path):
            pytest.skip("Playwright IFC file not found")

        result = validate_ifc_file(ifc_path)
        assert result is True, "Validation should pass"
